File: 0784-insert-into-a-binary-search-tree/0784-insert-into-a-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* insertIntoBST(struct TreeNode* root, int val) {

    if(!root){
        struct TreeNode* newNode = malloc(sizeof(struct TreeNode));
        newNode->val = val; 
        newNode->left = NULL;
        newNode->right = NULL;
        root = newNode;
        return root;
    }

    struct TreeNode* copy = root;
    struct TreeNode* pre = NULL;

    while(copy){
        pre = copy;
        if(val <= copy->val){
            copy = copy->left;
        } else {
            copy = copy->right;
        }
    }

    if(val > pre->val) {
        struct TreeNode* newN = malloc(sizeof(struct TreeNode));
        newN->val = val; 
        newN->left = NULL;
        newN->right = NULL;
        pre->right = newN;
    } else {
        struct TreeNode* newN = malloc(sizeof(struct TreeNode));
        newN->val = val; 
        newN->left = NULL;
        newN->right = NULL;
        pre->left = newN;
    }

    return root;
    
}

File: 0784-insert-into-a-binary-search-tree/0784-insert-into-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {

        if(root == null){
            root = new TreeNode(val);
            return root;
        }

        TreeNode copy = root;
        TreeNode pre = null;

        while(copy != null){
            pre = copy;
            if(val <= copy.val){
                copy = copy.left;
            } else {
                copy = copy.right;
            }
        }

        if(val > pre.val){
            pre.right = new TreeNode(val);
        }else {
            pre.left = new TreeNode(val);
        }

        return root;
        
    }
}